src/v/kafka/server/group_manager.cc

```diff
@@ -618,6 +629,91 @@ group_manager::txn_offset_commit(txn_offset_commit_request&& r) {
       .finally([this] { _catchup_lock.read_unlock(); });
 }
 
+ss::future<cluster::begin_group_tx_reply>
+group_manager::begin_tx(cluster::begin_group_tx_request&& r) {
+    if (!_catchup_lock.try_read_lock()) {
+        // transaction operations can't run in parallel with loading
+        // state from the log (happens once per term change)
+        vlog(klog.warn, "can't process a tx: coordinator_load_in_progress");
+        return ss::make_ready_future<cluster::begin_group_tx_reply>(
+          make_begin_tx_reply(cluster::tx_errc::coordinator_load_in_progress));
+    }
+
+    return do_begin_tx(std::move(r)).finally([this] {
+        _catchup_lock.read_unlock();
+    });
+}
+
+ss::future<cluster::begin_group_tx_reply>
+group_manager::do_begin_tx(cluster::begin_group_tx_request&& r) {
+    auto error = validate_group_status(
+      r.ntp, r.group_id, offset_commit_api::key);
+    if (error != error_code::none) {
+        if (error == error_code::not_coordinator) {
+            co_return make_begin_tx_reply(cluster::tx_errc::not_coordinator);
+        } else {
+            co_return make_begin_tx_reply(cluster::tx_errc::timeout);
+        }
+    }
+
+    auto partition = _partitions.find(r.ntp)->second;
+
+    if (partition->partition->term() != partition->term) {
+        co_return make_begin_tx_reply(cluster::tx_errc::not_coordinator);
+    }
+
+    auto fence_it = partition->fence_pid_epoch.find(r.pid.get_id());
+    if (
+      fence_it == partition->fence_pid_epoch.end()
+      || r.pid.get_epoch() > fence_it->second) {
+        // TODO: https://app.clubhouse.io/vectorized/story/2200
+        // include producer_id into key to make it unique-ish
+        // to prevent being GCed by the compaction
+        auto batch = cluster::make_fence_batch(
+          fence_control_record_version, r.pid);
+        auto reader = model::make_memory_record_batch_reader(std::move(batch));
+        auto e = co_await partition->partition->replicate(
+          partition->term,
+          std::move(reader),
+          raft::replicate_options(raft::consistency_level::quorum_ack));
+
+        if (!e) {
+            vlog(
+              klog.error,
+              "Error \"{}\" on replicating pid:{} fencing batch",
+              e.error(),
+              r.pid);
+            co_return make_begin_tx_reply(cluster::tx_errc::timeout);
+        }
+
+        fence_it = partition->fence_pid_epoch.find(r.pid.get_id());
```

Comment:
there is another part of the condition
> fence_it == partition->fence_pid_epoch.end() || **r.pid.get_epoch() > fence_it->second**

Reply:
mm, I think my comment wasn't clear. since the look-up done before this block of code is repeated here before leaving the block the implication would seem to be that something happens that might alter the result of the lookup. but the only thing in the block that happens is a batch is made and replicated. I think a comment about why this lookup is repeated is critical to understanding what is happening here.

Reply:
right, this is a concurrency issue. I have to recheck after the replication ended because while we were waiting there might be another begin_tx call with a higher pid which fenced out the current invocation

Reply:
perfect. that's a good thing to drop in as a comment.

Reply:
updated